File: 2016/AdventOfCSharp/AdventOfCSharp/Day1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AdventOfCSharp
{
    internal static class Day1
    {
        internal static void Solve2(string input)
        {
            var (curX, curY) = (0, 0);
            var (dirX, dirY) = (0, 1);
            HashSet<(int, int)> seen = new HashSet<(int, int)>();
            foreach (string s in input.Split(new[] { ", " }, StringSplitOptions.RemoveEmptyEntries))
            {
                if (s[0] == 'R')
                    (dirX, dirY) = (dirY, -dirX);
                else
                    (dirX, dirY) = (-dirY, dirX);

                int distance = int.Parse(s.Substring(1));
                for (int i = 0; i < distance; i++)
                {
                    if (!seen.Add((curX, curY)))
                    {
                        Console.WriteLine("Distance: {0}", Math.Abs(curX) + Math.Abs(curY));
                        goto Done;
                    }

                    (curX, curY) = (curX + dirX, curY + dirY);
                }
            }

            Done:
            ;
        }
    }
}

File: 2016/AdventOfCode/AdventOfCode/Day1.fs
module Day1
open System

let solve (input : string) = 
    let turns = input.Split ([| ", " |], StringSplitOptions.None)
    let applyTurn (curX, curY, dirX, dirY, visited : Set<int * int>) (turn : string) =
        if Set.contains (curX, curY) visited then
            (curX, curY, dirX, dirY, visited)
        else
            let (dirX, dirY) =
                if turn.[0] = 'R' then
                    (dirY, -dirX)
                else
                    (-dirY, dirX)

            let distance = int turn.[1..]
            let (newX, newY) = (curX + dirX * distance, curY + dirY * distance)
            (newX, newY, dirX, dirY, Set.add (curX, curY) visited)

    let visited : Set<int * int> = Set.empty
    let (endX, endY, _, _, _) = Seq.fold applyTurn (0, 0, 0, 1, visited) turns
    printfn "Distance: %O" (abs(endX) + abs(endY))